baseapp/abci.go

```diff
@@ -238,19 +241,52 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // work in a block before proposing it.
 //
 // Transactions can be modified, removed, or added by the application. Since the
-// application maintains it's own local mempool, it will ignore the transactions
+// application maintains its own local mempool, it will ignore the transactions
 // provided to it in RequestPrepareProposal. Instead, it will determine which
 // transactions to return based on the mempool's semantics and the MaxTxBytes
 // provided by the client's request.
 //
-// Note, there is no need to execute the transactions for validity as they have
-// already passed CheckTx.
-//
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	// TODO: Implement.
-	return abci.ResponsePrepareProposal{Txs: req.Txs}
+	var (
+		txsBytes  [][]byte
+		byteCount int64
+	)
+
+	ctx := app.getContextForTx(runTxPrepareProposal, []byte{})
+	iterator := app.mempool.Select(ctx, req.Txs)
+
+	for iterator != nil {
+		memTx := iterator.Tx()
+
+		bz, err := app.txEncoder(memTx)
+		if err != nil {
+			panic(err)
+		}
+
+		txSize := int64(len(bz))
+
+		// NOTE: runTx was already run in CheckTx which calls mempool.Insert so ideally everything in the pool
+		// should be valid. But some mempool implementations may insert invalid txs, so we check again.
+		_, _, _, _, err = app.runTx(runTxPrepareProposal, bz)
+		if err != nil {
+			err := app.mempool.Remove(memTx)
+			if err != nil && !errors.Is(err, mempool.ErrTxNotFound) {
+				panic(err)
+			}
+			iterator = iterator.Next()
+			continue
+		} else if byteCount += txSize; byteCount <= req.MaxTxBytes {
+			txsBytes = append(txsBytes, bz)
+		} else {
+			break
```

Comment:
Food for thought:

This `else` block seems to suggest that the txs queued up in the iterator won't be part of the returned response once the `MaxTxBytes` has been reached. Not processing all the queued up txs is not desirable in certain app scenarios.

For example, let's say `Select` returns txs [A1, A2, B1, B2] but `PrepareProposal` only processes [A1, A2] and leaves out [B1, B2], because the `MaxTxBytes` has been reached. `Select` returned the txs in the above order, because the app has an invariant that txs of type A must appear before txs of type B, but from the app's perspective, txs of type B are much more important to include in the block than type A. So in this case, the invariant is respected, but the app doesn't get to include txs of type B and is not alerted about this behavior.

Thinking out loud, to prevent this issue, the app can validate that the txs returned in `Select` is under the `MaxTxBytes` cap. However, the app doesn't get to learn if all the txs returned in `Select` were processed or not (it speculates that the txs were processed by `PrepareProposal`, but there's no explicit signal).

Thinking out loud, some suggestions:
* Hint `MaxTxBytes` as part of `Select`
* Add a way for the app to learn that not all txs returned in `Select` were processed by `PrepareProposal` and retry `Select` / `PrepareProposal`
* Add a mode where "process all txs returned by `Select` or fail"

Reply:
The reason we don't pass MaxBytes to `Select` is that we need to ensure that selected txs are valid.

Reply:
We discussed this further offline. https://github.com/cosmos/cosmos-sdk/pull/13831 fixes the concern.